src/unix-manager.c

```diff
@@ -700,18 +702,112 @@ TmEcode UnixManagerCaptureModeCommand(json_t *cmd,
     SCReturnInt(TM_ECODE_OK);
 }
 
-TmEcode UnixManagerReloadRules(json_t *cmd, json_t *server_msg, void *data)
+TmEcode UnixManagerReloadRulesWrapper(json_t *cmd, json_t *server_msg, void *data, int wait)
 {
     SCEnter();
-    DetectEngineReloadStart();
+    int r = DetectEngineReloadStart();
 
-    while (DetectEngineReloadIsDone() == 0)
-        usleep(100);
+    if (wait) {
+        while (DetectEngineReloadIsDone() == 0)
+            usleep(100);
+    } else {
+        if (DetectEngineReloadIsDone() == 0 && r == -1) {
```

Comment:
On 12-04-16 14:33, Giuseppe Longo wrote:
> In src/unix-manager.c
> <https://github.com/inliniac/suricata/pull/1998#discussion_r59367085>:
> 
>>  
>> -    while (DetectEngineReloadIsDone() == 0)
>> -        usleep(100);
>> +    if (wait) {
>> +        while (DetectEngineReloadIsDone() == 0)
>> +            usleep(100);
>> +    } else {
>> +        if (DetectEngineReloadIsDone() == 0 && r == -1) {
> 
> When the reload is completed, you can't restart a new one again because
> detect_sync.state is not set IDLE.
> To see that try to remove DetectEngineReloadIsDone() from the else
> statement, so it should be called.

This function calls DetectEngineReloadStart(), which returns -1. This
means it didn't touch the detect_sync.state which was already set to RELOAD.

wait == 0, so we're not waiting. Why are we not just returning an error
message?

If the detect_sync.state is wrong somehow, that should be fixed in the
proper place, not here.